NebulaModel/Utils/NativeInterop.cs

```diff
@@ -17,5 +19,50 @@ public static void HideWindow()
         {
             ShowWindow(GetActiveWindow(), SW_HIDE);
         }
+
+
+        [DllImport("Kernel32")]
+        private static extern bool SetConsoleCtrlHandler(EventHandler handler, bool add);
+
+        private delegate bool EventHandler(CtrlType sig);
+
+        private enum CtrlType
+        {
+            CTRL_C_EVENT = 0,
+            CTRL_BREAK_EVENT = 1,
+            CTRL_CLOSE_EVENT = 2,
+            CTRL_LOGOFF_EVENT = 5,
+            CTRL_SHUTDOWN_EVENT = 6
+        }
+
+        private static bool Handler(CtrlType sig)
+        {
+            Console.WriteLine($"Exiting app due to {sig}");
+
+            switch(sig)
+            {
+                // Only CTRL_C, CTRL_BREAK events have no timeout
+                case CtrlType.CTRL_C_EVENT:
+                case CtrlType.CTRL_BREAK_EVENT:
+                    Console.WriteLine($"Start saving to last exit...");
+                    ManualResetEvent mre = new(false);
+                    ThreadingHelper.Instance.StartSyncInvoke(() => {
+                        UIRoot.instance.uiGame.escMenu.OnButton6Click(); //ESC Menu - ExitProgram;
+                        mre.Set();
+                    });
+                    mre.WaitOne();
+                    Console.WriteLine($"Saving completed!");
+                    Thread.Sleep(1000);
+                    return false;
+            }
+            Thread.Sleep(500);
+            return false;
+        }
+
+        public static void SetConsoleCtrlHandler()
+        {
+            bool result = SetConsoleCtrlHandler(new EventHandler(Handler), true);
```

Comment:
Even though it's a static method that's being used as the callback, maybe a comment can be added to signal that if the handler is no longer static, it'll can get GC'd because nothing is keeping a reference to the delegate.